core/sr-primitives/src/traits.rs

```diff
@@ -1055,21 +1055,31 @@ pub trait OpaqueKeys: Clone {
 	/// Get the raw bytes of key with key-type ID `i`.
 	fn get_raw(&self, i: super::KeyTypeId) -> &[u8];
 	/// Get the decoded key with index `i`.
-	fn get<T: Decode>(&self, i: super::KeyTypeId) -> Option<T> { T::decode(&mut self.get_raw(i)) }
+	fn get<T: Decode>(&self, i: super::KeyTypeId) -> Option<T> {
+		T::decode(&mut self.get_raw(i)).ok()
+	}
 	/// Verify a proof of ownership for the keys.
 	fn ownership_proof_is_valid(&self, _proof: &[u8]) -> bool { true }
 }
 
 struct TrailingZeroInput<'a>(&'a [u8]);
 impl<'a> codec::Input for TrailingZeroInput<'a> {
-	fn read(&mut self, into: &mut [u8]) -> usize {
+	fn remaining_len(&mut self) -> Result<usize, codec::Error> {
+		// TODO TODO: this might not be super correct: remaining_len change during execution
+		// Also this gives attack possibility no ?
```

Comment:
still to be thought: this TrailingZeroInput is adding infinite number of 0 after input. but input length is now restricted to be known and is used to constrain allocation. thus I'm not the how we should implement it for now